From b5bdd29596665e694b46071de3d9d6221dedc0df Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <alexey.makhalov@broadcom.com>
Date: Tue, 16 Sep 2025 20:44:35 +0000
Subject: [PATCH 16/16] Move crypto_inc to lib/crypto/utils.c

crypto_inc is a utility function similar to crypto_xor.
algapi.c is a wrong place for it.

Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
---
 crypto/algapi.c                | 31 -------------------------------
 include/crypto/algapi.h        |  2 --
 include/crypto/utils.h         |  2 ++
 lib/crypto/utils.c             | 30 ++++++++++++++++++++++++++++++
 4 files changed, 32 insertions(+), 33 deletions(-)

diff --git a/crypto/algapi.c b/crypto/algapi.c
index 685d1930d..d4fc7cc52 100644
--- a/crypto/algapi.c
+++ b/crypto/algapi.c
@@ -963,37 +963,6 @@ struct crypto_async_request *crypto_dequeue_request(struct crypto_queue *queue)
 }
 EXPORT_SYMBOL_GPL(crypto_dequeue_request);
 
-static inline void crypto_inc_byte(u8 *a, unsigned int size)
-{
-	u8 *b = (a + size);
-	u8 c;
-
-	for (; size; size--) {
-		c = *--b + 1;
-		*b = c;
-		if (c)
-			break;
-	}
-}
-
-void crypto_inc(u8 *a, unsigned int size)
-{
-	__be32 *b = (__be32 *)(a + size);
-	u32 c;
-
-	if (IS_ENABLED(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS) ||
-	    IS_ALIGNED((unsigned long)b, __alignof__(*b)))
-		for (; size >= 4; size -= 4) {
-			c = be32_to_cpu(*--b) + 1;
-			*b = cpu_to_be32(c);
-			if (likely(c))
-				return;
-		}
-
-	crypto_inc_byte(a, size);
-}
-EXPORT_SYMBOL_GPL(crypto_inc);
-
 unsigned int crypto_alg_extsize(struct crypto_alg *alg)
 {
 	return alg->cra_ctxsize +
diff --git a/include/crypto/algapi.h b/include/crypto/algapi.h
index b66ba99f7..aee693fc3 100644
--- a/include/crypto/algapi.h
+++ b/include/crypto/algapi.h
@@ -182,8 +182,6 @@ static inline unsigned int crypto_queue_len(struct crypto_queue *queue)
 	return queue->qlen;
 }
 
-void crypto_inc(u8 *a, unsigned int size);
-
 static inline void *crypto_tfm_ctx(struct crypto_tfm *tfm)
 {
 	return tfm->__crt_ctx;
diff --git a/include/crypto/utils.h b/include/crypto/utils.h
index 2594f4577..011b2e6b5 100644
--- a/include/crypto/utils.h
+++ b/include/crypto/utils.h
@@ -12,6 +12,8 @@
 #include <linux/types.h>
 
 void __crypto_xor(u8 *dst, const u8 *src1, const u8 *src2, unsigned int size);
+void crypto_inc(u8 *a, unsigned int size);
+
 
 static inline void crypto_xor(u8 *dst, const u8 *src, unsigned int size)
 {
diff --git a/lib/crypto/utils.c b/lib/crypto/utils.c
index d579abf17..deea0ffb1 100644
--- a/lib/crypto/utils.c
+++ b/lib/crypto/utils.c
@@ -8,6 +8,7 @@
 #include <linux/unaligned.h>
 #include <crypto/utils.h>
 #include <linux/bitops.h>
+#include <linux/align.h>
 
 /*
  * XOR @len bytes from @src1 and @src2 together, writing the result to @dst
@@ -84,3 +85,32 @@ void __crypto_xor(u8 *dst, const u8 *src1, const u8 *src2, unsigned int len)
 		*dst++ = *src1++ ^ *src2++;
 }
 
+static inline void crypto_inc_byte(u8 *a, unsigned int size)
+{
+	u8 *b = (a + size);
+	u8 c;
+
+	for (; size; size--) {
+		c = *--b + 1;
+		*b = c;
+		if (c)
+			break;
+	}
+}
+
+void crypto_inc(u8 *a, unsigned int size)
+{
+	__be32 *b = (__be32 *)(a + size);
+	u32 c;
+
+	if (IS_ENABLED(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS) ||
+	    IS_ALIGNED((unsigned long)b, __alignof__(*b)))
+		for (; size >= 4; size -= 4) {
+			c = be32_to_cpu(*--b) + 1;
+			*b = cpu_to_be32(c);
+			if (likely(c))
+				return;
+		}
+
+	crypto_inc_byte(a, size);
+}
-- 
2.40.4

